worlds/messenger/__init__.py

```diff
@@ -54,58 +67,139 @@ class MessengerWorld(World):
+
+        self.multiworld.early_items[self.player]["Meditation"] = self.options.early_meditation.value
+
         self.shop_prices, self.figurine_prices = shuffle_shop_prices(self)
 
+        starting_portals = ["Autumn Hills", "Howling Grotto", "Glacial Peak", "Riviere Turquoise", "Sunken Shrine", "Searing Crags"]
+        self.starting_portals = [f"{portal} Portal"
+                                 for portal in starting_portals[:3] +
+                                 self.random.sample(starting_portals[3:], k=self.options.available_portals - 3)]
+        self.plando_portals = []
+        self.portal_mapping = []
+        self.spoiler_portal_mapping = {}
+        self.transitions = []
+
+    @classmethod
+    def stage_generate_early(cls, multiworld: MultiWorld):
+        if multiworld.players > 1:
+            return
+        cls.generate_output = generate_output
+
     def create_regions(self) -> None:
         # MessengerRegion adds itself to the multiworld
-        for region in [MessengerRegion(reg_name, self) for reg_name in REGIONS]:
-            if region.name in REGION_CONNECTIONS:
-                region.add_exits(REGION_CONNECTIONS[region.name])
+        # create simple regions
+        for level in LEVELS:
+            MessengerRegion(level, self)
+        # create and connect complex regions that have sub-regions
+        for region in [MessengerRegion(f"{parent} - {reg_name}", self, parent)
+                       for parent, sub_region in CONNECTIONS.items()
+                       for reg_name in sub_region]:
+            region_name = region.name.replace(f"{region.parent} - ", "")
+            connection_data = CONNECTIONS[region.parent][region_name]
+            for exit_region in connection_data:
+                region.connect(self.multiworld.get_region(exit_region, self.player))
+        # all regions need to be created before i can do these connections so we create and connect the complex first
+        for region_name in [level for level in LEVELS if level in REGION_CONNECTIONS]:
+            region = self.multiworld.get_region(region_name, self.player)
+            region.add_exits(REGION_CONNECTIONS[region.name])
 
     def create_items(self) -> None:
         # create items that are always in the item pool
+        main_movement_items = ["Rope Dart", "Wingsuit"]
         itempool: List[MessengerItem] = [
             self.create_item(item)
             for item in self.item_name_to_id
-            if item not in
-            {
-                "Power Seal", *NOTES, *FIGURINES,
-                *{collected_item.name for collected_item in self.multiworld.precollected_items[self.player]},
-            } and "Time Shard" not in item
+            if item not in {
+                   "Power Seal", *NOTES, *FIGURINES, *main_movement_items,
+                   *{collected_item.name for collected_item in self.multiworld.precollected_items[self.player]},
+               } and "Time Shard" not in item
```

Comment:
this is a tiny bit slower, but the real reason i didn't do this is because i thought i'd need to move all the other items further to the right making the comprehension unreadable. IDE seems to want to move them further to the left though so using it.